diff --git a/megatron/training/initialize.py b/megatron/training/initialize.py
index a645f89..f834616 100644
--- a/megatron/training/initialize.py
+++ b/megatron/training/initialize.py
@@ -108,7 +108,7 @@ def _compile_dependencies():
     # Compile dataset C++ code.
     # =========================
     # TODO: move this to ninja
-    if torch.distributed.get_rank() == 0:
+    if torch.distributed.get_rank() % args.tensor_model_parallel_size == 0:
         start_time = time.time()
         print("> compiling dataset index builder ...")
         from megatron.core.datasets.utils import compile_helpers
